### PR TITLE
[TLS 1.3] Client Authentication in the main Handshake

### PR DESCRIPTION
**Pull request by hrantzsch**
### Pull Request Dependencies

Before merging this, we should first review merge:

* https://github.com/randombit/botan/pull/2968
* https://github.com/randombit/botan/pull/2922

### Description

This implements:
 * client authentication during the main handshake
 * coalescing of multiple handshake messages (which is required for the RFC 8448 client auth test) into a single (encrypted) record

Post-handshake authentication is left for future work. Rationale: A server may always ask for client authentication during the handshake. Post-handshake auth can be disabled by the client, by not negotiating the "post_handshake_auth" extension in the Client Hello.

#### TODO

* ~We should explicitly support the "signature_algorithm_cert" extension~
  The client can choose to not use this extension. Hence its support becomes crucial when implementing the server only.

### CODE REVIEW

File: src/lib/tls/tls13/tls_client_impl_13.cpp [hunk 1]
```diff
@@ -412,9 +428,52 @@ void Client_Impl_13::handle(const Finished_13& finished_msg)
...
+
+      std::vector<X509_Certificate> client_certs = credentials_manager().find_cert_chain(
+            peer_allowed_signature_algos,
+            cert_request.acceptable_CAs(),
+            "tls-client",
+            m_info.hostname());
+
+      flight.add(m_handshake_state.sending(Certificate_13(std::move(client_certs))));
+
+      // TODO: empty certs are allowed if no certificate is found -- don't fail
```
Comment (hrantzsch) [inline src/lib/tls/tls13/tls_client_impl_13.cpp:450]: don't forget about me

File: src/lib/tls/msg_cert_verify.cpp [hunk 2]
```diff
@@ -8,6 +8,7 @@
 * Botan is released under the Simplified BSD License (see license.txt)
 */
 
+#include "botan/internal/stl_util.h"
```
Comment (reneme) [inline src/lib/tls/msg_cert_verify.cpp:11]: Use `#include <...>`

**Comment by reneme** [comment]
Rebased and retargeted to master.

File: src/lib/tls/msg_cert_verify.cpp [hunk 3]
```diff
@@ -106,8 +107,73 @@ bool Certificate_Verify_12::verify(const X509_Certificate& cert,
 
 #if defined(BOTAN_HAS_TLS_13)
 
+namespace {
+
+std::vector<uint8_t> message(Connection_Side side, const Transcript_Hash& hash)
+   {
+   std::vector<uint8_t> msg(64, 0x20);
+   msg.reserve(64 + 32 + 1 + hash.size());
```
[the remark this answers is not in the document]
Comment (reneme) [reply src/lib/tls/msg_cert_verify.cpp:115]: Yes (⊙_⊙;)

File: src/lib/tls/tls13/msg_certificate_req_13.cpp [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+      TLSEXT_CERTIFICATE_AUTHORITIES,
+      // TLSEXT_OID_FILTERS,                   // NYI
+      TLSEXT_SIGNATURE_ALGORITHMS_CERT,
+      };
+
+   if(m_extensions.contains_implemented_extensions_other_than(allowed_extensions))
+      {
+      throw TLS_Exception(Alert::UNSUPPORTED_EXTENSION,
+                          "Certificate Request contained an extension that is not allowed");
+      }
```
[the remark this answers is not in the document]
Comment (reneme) [reply src/lib/tls/tls13/msg_certificate_req_13.cpp:59]: Both is true. In [RFC 8446 4.2 there is a table](https://www.rfc-editor.org/rfc/rfc8446#page-37) which declares the valid extensions for each handshake message. For "Certificate Request", that are the six extensions listed in `allowed_extensions`. I added another RFC reference comment to make this clearer.

On top of those, there might be more (unknown) extensions. Those shall be ignored. And that's what `contains_implemented_extensions_other_than()` is supposed to do: Reject extensions that we know about but that are not listed in `allowed_extensions`.

Also RFC 8446 4.2 states that we should report an illegal_parameter alert in that case (in-lieu of unsupported_extension). I updated that accordingly: hopefully BoGo doesn't mind.

**Review by lieser — APPROVED** [review]

**Pull request merged**